closed/adds/jdk/src/share/classes/sun/security/ec/NativeECDHKeyAgreement.java

```diff
@@ -0,0 +1,264 @@
+            return this.javaImplementation.engineDoPhase(key, lastPhase);
+        }
+        if (this.privateKey == null) {
+            throw new IllegalStateException("Not initialized");
+        }
+        if (this.publicKey != null) {
+            throw new IllegalStateException("Phase already executed");
+        }
+        if (!lastPhase) {
+            throw new IllegalStateException
+                ("Only two party agreement supported, lastPhase must be true");
+        }
+        if (!(key instanceof PublicKey)) {
+            throw new InvalidKeyException
+                ("Key must be an instance of PublicKey");
+        }
+        /* attempt to translate the key if it is not an ECKey */
+        this.publicKey = (ECPublicKeyImpl) ECKeyFactory.toECKey(key);
+
+        ECParameterSpec params = this.publicKey.getParams();
+        int keyLenBits = params.getCurve().getField().getFieldSize();
+        this.secretLen = (keyLenBits + 7) >> 3;
+
+        return null;
+    }
+
+    @Override
+    protected byte[] engineGenerateSecret() throws IllegalStateException {
+        if (this.javaImplementation != null) {
+            return this.javaImplementation.engineGenerateSecret();
+        }
+        byte[] secret = new byte[this.secretLen];
+        try {
+            engineGenerateSecret(secret, 0);
+        } catch (ShortBufferException e) {
+            /* should not happen */
+        }
+        return secret;
+    }
+
+    @Override
+    protected int engineGenerateSecret(byte[] sharedSecret, int offset)
+            throws IllegalStateException, ShortBufferException {
+        if (this.javaImplementation != null) {
+            return this.javaImplementation.engineGenerateSecret(sharedSecret, offset);
+        }
+        if ((offset + this.secretLen) > sharedSecret.length) {
+            throw new ShortBufferException("Need " + this.secretLen
+                + " bytes, only " + (sharedSecret.length - offset)
+                + " available");
+        }
+        if ((this.privateKey == null) || (this.publicKey == null)) {
+            throw new IllegalStateException("Not initialized correctly");
+        }
+        long nativePublicKey = this.publicKey.getNativePtr();
+        long nativePrivateKey = this.privateKey.getNativePtr();
+        if ((nativePublicKey == -1) || (nativePrivateKey == -1)) {
+            if (curveSupported.containsKey(this.curve)) {
+                throw new ProviderException("Could not convert keys to native format");
+            } else {
+                curveSupported.put(this.curve, false);
+                try {
+                    this.useJavaImplementation(this.privateKey, null, this.curve);
+                    this.javaImplementation.engineDoPhase(this.publicKey, true);
+                } catch (InvalidKeyException e) {
+                    System.out.println(e.toString());
+                }
+                return this.javaImplementation.engineGenerateSecret(sharedSecret, offset);
+            }
```

Comment:
When do you anticipate this will happen?
If this code is to stay, it needs a better response to `InvalidKeyException`.

Reply:
It should not happen, since `this.privateKey` is an instance of `ECPrivateKey` and `this.publicKey` is an instance of `ECPublicKey`, `InvalidKeyException` should not be thrown.